esphome/components/esp32_ble_server/__init__.py

```diff
@@ -11,44 +19,337 @@
+)
+
+CHARACTERISTIC_VALUE_SCHEMA = cv.Any(
+    cv.boolean,
+    cv.float_,
+    cv.hex_uint8_t,
+    cv.hex_uint16_t,
+    cv.hex_uint32_t,
+    cv.int_,
+    cv.templatable(cv.All(cv.ensure_list(cv.hex_uint8_t), cv.Length(min=1))),
+    cv.string,
+)
+
+DESCRIPTOR_SCHEMA = cv.Schema(
+    {
+        cv.GenerateID(): cv.declare_id(BLEDescriptor),
+        cv.Required(CONF_UUID): UUID_SCHEMA,
+        cv.Required(CONF_VALUE): DESCRIPTOR_VALUE_SCHEMA,
+    }
+)
+
+SERVICE_CHARACTERISTIC_SCHEMA = cv.Schema(
+    {
+        cv.GenerateID(): cv.declare_id(BLECharacteristic),
+        cv.Required(CONF_UUID): UUID_SCHEMA,
+        cv.Optional(CONF_READ, default=False): cv.boolean,
+        cv.Optional(CONF_WRITE, default=False): cv.boolean,
+        cv.Optional(CONF_NOTIFY, default=False): cv.boolean,
+        cv.Optional(CONF_BROADCAST, default=False): cv.boolean,
+        cv.Optional(CONF_INDICATE, default=False): cv.boolean,
+        cv.Optional(CONF_WRITE_NO_RESPONSE, default=False): cv.boolean,
+        cv.Optional(CONF_VALUE): CHARACTERISTIC_VALUE_SCHEMA,
+        cv.GenerateID(CONF_VALUE_ACTION_ID_): cv.declare_id(
+            BLECharacteristicSetValueAction
+        ),
+        cv.Optional(CONF_DESCRIPTORS, default=[]): cv.ensure_list(DESCRIPTOR_SCHEMA),
+        cv.Optional(CONF_ON_WRITE): automation.validate_automation(
+            {cv.GenerateID(): cv.declare_id(BLECharacteristic)}, single=True
+        ),
+    }
+)
+
+SERVICE_SCHEMA = cv.Schema(
+    {
+        cv.GenerateID(): cv.declare_id(BLEService),
+        cv.Required(CONF_UUID): UUID_SCHEMA,
+        cv.Optional(CONF_ADVERTISE, default=False): cv.boolean,
+        cv.Optional(CONF_NUM_HANDLES, default=0): cv.int_,
+        cv.Optional(CONF_CHARACTERISTICS, default=[]): cv.ensure_list(
+            SERVICE_CHARACTERISTIC_SCHEMA
+        ),
+    }
+)
+
 CONFIG_SCHEMA = cv.Schema(
     {
         cv.GenerateID(): cv.declare_id(BLEServer),
         cv.GenerateID(esp32_ble.CONF_BLE_ID): cv.use_id(esp32_ble.ESP32BLE),
         cv.Optional(CONF_MANUFACTURER, default="ESPHome"): cv.string,
         cv.Optional(CONF_MANUFACTURER_DATA): cv.Schema([cv.hex_uint8_t]),
         cv.Optional(CONF_MODEL): cv.string,
+        cv.Optional(CONF_SERVICES, default=[]): cv.ensure_list(SERVICE_SCHEMA),
     }
 ).extend(cv.COMPONENT_SCHEMA)
 
 
+def parse_properties(char_conf):
+    result = 0
+    if char_conf[CONF_READ]:
+        result = result | BLECharacteristic_ns.PROPERTY_READ
+    if char_conf[CONF_WRITE]:
+        result = result | BLECharacteristic_ns.PROPERTY_WRITE
+    if char_conf[CONF_NOTIFY]:
+        result = result | BLECharacteristic_ns.PROPERTY_NOTIFY
+    if char_conf[CONF_BROADCAST]:
+        result = result | BLECharacteristic_ns.PROPERTY_BROADCAST
+    if char_conf[CONF_INDICATE]:
+        result = result | BLECharacteristic_ns.PROPERTY_INDICATE
+    if char_conf[CONF_WRITE_NO_RESPONSE]:
+        result = result | BLECharacteristic_ns.PROPERTY_WRITE_NR
+    return result
```

Comment:
```suggestion
PROPERTY_MAP = {
    CONF_READ: BLECharacteristic_ns.PROPERTY_READ,
    CONF_WRITE: BLECharacteristic_ns.PROPERTY_WRITE,
    CONF_NOTIFY: BLECharacteristic_ns.PROPERTY_NOTIFY,
    CONF_BROADCAST: BLECharacteristic_ns.PROPERTY_BROADCAST,
    CONF_INDICATE: BLECharacteristic_ns.PROPERTY_INDICATE,
    CONF_WRITE_NO_RESPONSE: BLECharacteristic_ns.PROPERTY_WRITE_NR,
}

def parse_properties(char_conf):
    return sum([PROPERTY_MAP[k] for k in char_conf if k in PROPERTY_MAP], start=0)
```

Can then also use the map keys to generate the corresponding entries in the config schema e.g.
```
).extend({
    cv.Optional(k, default=False): cv.boolean for k in PROPERTY_MAP.keys()
})
```

Reply:
You are right! Thanks! I have included this change in 96026e8d07500f9f2b5db336b546488dd8c20f52